File: cpp/lc543.cxx
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
public:
    int m_dep;
    int diameterOfBinaryTree(TreeNode* root) {
        if (root) {
            int r_max = max_depth(root->right);
            int l_max = max_depth(root->left);
            return max(r_max+l_max, m_dep);
        } else {
            return 0;
        }
    }

    int max_depth(TreeNode* root) {
        if (root) {
            int r_max = max_depth(root->right);
            int l_max = max_depth(root->left);
            m_dep = max(m_dep, r_max + l_max);
            return 1 + max(r_max, l_max);
        } else {
            return 0;
        }
    }
};

File: src/bin/lc543.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}
//
impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}
struct Solution;
fn main() {}
// @lc code=start
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(node) = root {
            1 + Self::max_depth(&node.borrow().left.clone()) + Self::max_depth(&node.borrow().right.clone())
        } else {
            0
        }
    }

    fn max_depth(root: &Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(node) = root {
            let r_max = Self::max_depth(&node.borrow().right);
            let l_max = Self::max_depth(&node.borrow().left);
            std::cmp::max(1 + r_max, 1 + l_max)
        } else {
            0
        }
    }
}
// @lc code=end